File: backend/app/schemas/api_keys.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from backend.app.models.enums import ApiKeyStatus


class ApiKeyPublic(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: int
    name: str
    key_prefix: str
    scopes: list[str]
    status: ApiKeyStatus
    expires_at: datetime | None
    last_used_at: datetime | None
    created_at: datetime
    updated_at: datetime


class ApiKeyCreate(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    scopes: list[str] = Field(default_factory=list)
    expires_at: datetime | None = None
    user_id: int | None = None


class ApiKeyUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=120)
    scopes: list[str] | None = None
    status: ApiKeyStatus | None = None
    expires_at: datetime | None = None


class ApiKeyCreateResponse(BaseModel):
    key: ApiKeyPublic
    api_key: str

File: frontend/src/api/apiKeys.ts
import { http } from './http'

export type ApiKeyStatus = 'active' | 'disabled' | 'expired'

export interface ApiKey {
  id: number
  user_id: number
  name: string
  key_prefix: string
  scopes: string[]
  status: ApiKeyStatus
  expires_at: string | null
  last_used_at: string | null
  created_at: string
  updated_at: string
}

export interface ApiKeyCreate {
  name: string
  scopes?: string[]
  expires_at?: string | null
  user_id?: number | null
}

export interface ApiKeyCreateResponse {
  key: ApiKey
  api_key: string
}

export function fetchApiKeys() {
  return http.get<ApiKey[]>('/api/api-keys')
}

export function createApiKey(payload: ApiKeyCreate) {
  return http.post<ApiKeyCreateResponse>('/api/api-keys', payload)
}

export function disableApiKey(id: number) {
  return http.delete<void>(`/api/api-keys/${id}`)
}
